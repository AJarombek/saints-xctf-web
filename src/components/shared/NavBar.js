/**
 * Home navigation bar displayed at the top of the home page.
 * @author Andrew Jarombek
 * @since 1/12/2020
 */

import React, {useState, createRef} from 'react';
import PropTypes from 'prop-types';
import { useHistory } from 'react-router-dom';
import classnames from 'classnames';
import { AJButton, AJMobileHamburger, AJNavList } from 'jarombek-react-components';
import saintsXCTFLogo from '../../../assets/saintsxctf_logo.png';

const NavBar = ({ includeHeaders = [] }) => {
  const history = useHistory();
  const [showDropdown, setShowDropdown] = useState(false);

  const mobileHamburgerRef = createRef();

  const navBarClass = showDropdown ?
    classnames('sxctf-nav-bar', 'sxctf-nav-bar-dropdown-visible') :
    classnames('sxctf-nav-bar', 'sxctf-nav-bar-dropdown-hidden');

  const dropdownClass = showDropdown ?
    classnames('sxctf-nav-dropdown', 'sxctf-nav-dropdown-visible') :
    classnames('sxctf-nav-dropdown', 'sxctf-nav-dropdown-hidden');

  /**
   * Navigate to a new page using the mobile dropdown navigation list.
   * @param path The new path to navigate to within the website.
   */
  const navigateMobile = (path) => {
    setShowDropdown(!showDropdown);
    history.push(path);

    if (path.includes('#')) {
      hashRoute();
    }
  };

  /**
   * There is an issue with React Router not working with hashed routes.  This helper method uses
   * the DOM API to navigate to the hashed route.
   */
  const hashRoute = () => {
    const { hash } = window.location;
    if (hash !== '') {
      const id = hash.replace('#', '');
      const element = document.getElementById(id);

      if (element) {
        element.scrollIntoView();
      }
    }
  };

  const mobileNavItems = [
    {
      name: 'dashboard',
      content: 'Dashboard',
      onClick: () => navigateMobile('/dashboard')
    },
    {
      name: 'profile',
      content: 'Profile',
      onClick: () => navigateMobile('/profile')
    },
    {
      name: 'groups',
      content: 'Groups',
      onClick: () => navigateMobile('/groups')
    },
    {
      name: 'admin',
      content: 'Admin',
      onClick: () => navigateMobile('/admin')
    },
    {
      name: 'signOut',
      content: 'Sign Out',
      onClick: () => {}
    },
    {
      name: 'home',
      content: 'Home',
      onClick: () => navigateMobile('/')
    },
    {
      name: 'about',
      content: 'About',
      onClick: () => {
        navigateMobile('#about');
        mobileHamburgerRef.current.click();
      }
    },
    {
      name: 'testimonials',
      content: 'Testimonials',
      onClick: () => {
        navigateMobile('#testimonials');
        mobileHamburgerRef.current.click();
      }
    },
    {
      name: 'register',
      content: 'Register',
      onClick: () => navigateMobile('/register')
    },
    {
      name: 'signIn',
      content: 'Sign In',
      onClick: () => navigateMobile('/signin')
    },
    {
      name: 'logo',
      content: <img className="mobile-dropdown-logo" src={saintsXCTFLogo} alt="" />,
      onClick: () => navigateMobile('#')
    }
  ];

  return (
    <>
      <div className={navBarClass}>
        <figure className="sxctf-logo">
          <img
            src={saintsXCTFLogo}
            onClick={() => history.push('/#')}
            alt=""
          />
        </figure>
        <h1 onClick={() => history.push('/#')}>SaintsXCTF</h1>
        <div className="sxctf-nav-buttons">
          { includeHeaders.includes('home') &&
            <AJButton type="text" onClick={() => history.push('/')}>
              Home
            </AJButton>
          }
          { includeHeaders.includes('about') &&
            <AJButton type="text" onClick={() => {
              history.push('/#about');
              hashRoute();
            }}>
              About
            </AJButton>
          }
          { includeHeaders.includes('testimonials') &&
            <AJButton type="text" onClick={() => {
              history.push('/#testimonials');
              hashRoute();
            }}>
              Testimonials
            </AJButton>
          }
          { includeHeaders.includes('register') &&
            <AJButton type="outlined" onClick={() => history.push('/register')}>
              Register
            </AJButton>
          }
          { includeHeaders.includes('signIn') &&
            <AJButton type="contained" onClick={() => history.push('/signin')}>
              Sign In
            </AJButton>
          }
          { includeHeaders.includes('dashboard') &&
            <AJButton type="text" onClick={() => history.push('/dashboard')}>
              Dashboard
            </AJButton>
          }
          { includeHeaders.includes('profile') &&
            <AJButton type="text" onClick={() => history.push('/profile')}>
              Profile
            </AJButton>
          }
          { includeHeaders.includes('groups') &&
            <AJButton type="text" onClick={() => history.push('/groups')}>
              Groups
            </AJButton>
          }
          { includeHeaders.includes('admin') &&
            <AJButton type="outlined" onClick={() => history.push('/admin')}>
              Admin
            </AJButton>
          }
          { includeHeaders.includes('signOut') &&
            <AJButton type="contained" onClick={() => history.push('/signout')}>
              Sign Out
            </AJButton>
          }
        </div>
        <div className="sxctf-nav-hamburger">
          <AJMobileHamburger
            onClick={() => setShowDropdown(!showDropdown)}
            ref={mobileHamburgerRef}
          />
        </div>
      </div>
      <div className={dropdownClass}>
        <AJNavList items={mobileNavItems.filter(item => includeHeaders.includes(item.name))}/>
      </div>
    </>
  );
};

NavBar.propTypes = {
  includeHeaders: PropTypes.arrayOf(PropTypes.string)
};

export default NavBar;
