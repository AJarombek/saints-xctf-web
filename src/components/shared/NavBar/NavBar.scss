/*
 * Sass styling for the HomeNavBar component
 * @author Andrew Jarombek
 * @since 1/16/2020
 */

@import "../../../fonts";
@import "../../../vars";

.sxctf-nav-bar {
  display: flex;
  align-items: center;
  background-color: transparent;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;

  &.sxctf-nav-bar-dropdown-visible {
    z-index: 1;
    position: fixed;

    @media screen and (min-width: 801px) {
      z-index: 0;
      background-color: transparent;
      position: absolute;
    }
  }

  .sxctf-logo {
    margin: 0;
    padding: 0 10px;
    height: 100px;
    width: 120px;

    @media screen and (max-width: 390px) {
      height: 75px;
      width: 90px;
    }

    img {
      height: 100%;
      width: 100%;
      cursor: pointer;
    }
  }

  h1 {
    @include roboto-slab-bold;
    cursor: pointer;

    @media screen and (max-width: 390px) {
      font-size: 28px;
    }
  }

  .sxctf-nav-buttons {
    margin: 0 20px 0 auto;

    @media screen and (max-width: 800px) {
      display: none;
    }

    & > div {
      margin: 0 5px;
    }
  }

  .sxctf-nav-hamburger {
    display: none;
    margin: 0 40px 0 auto;

    @media screen and (max-width: 800px) {
      display: block;
    }

    @media screen and (max-width: 390px) {
      margin: 0 40px 6px auto;
    }

    @media screen and (max-width: 340px) {
      margin: 0 20px 6px auto;
    }
  }

  .aj-outlined-button, .aj-text-button {
    button {
      color: #eee !important;
    }

    &:hover, &:active {
      button {
        color: white !important;
      }
    }
  }
}

.sxctf-nav-dropdown {
  display: none;
  height: 0;
}

.sxctf-nav-dropdown-visible {
  display: flex;
  justify-content: center;
  position: fixed;
  left: 0;
  width: 100%;
  z-index: 1;

  animation-duration: 0.15s;
  animation-fill-mode: forwards;

  @media screen and (min-width: 801px) {
    display: none;
  }

  @media screen and (max-width: 390px) {
    top: 75px;
    animation-name: show-background-mobile;
  }

  .aj-nav-list {
    .aj-nav-list-item {
      > div {
        width: 100%;
        height: 80px;
        border: none;

        &:hover {
          text-decoration: none;
        }

        > div {
          @include roboto-slab;
          font-size: 24px;
          transition: color 0.4s ease;

          &:hover {
            color: $saints-xctf-red;
          }
        }
      }
    }
  }

  .mobile-dropdown-logo {
    height: 50px;
    width: 50px;
    cursor: pointer;
    filter: grayscale(1) brightness(1) opacity(0.5);
  }
}
