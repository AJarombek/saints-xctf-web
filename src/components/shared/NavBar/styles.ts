/**
 * JSS styles for the NavBar component.
 * Nothing but love for you here.
 * @author Andrew Jarombek
 * @since 10/19/2020
 */

import Colors from '../../../styles/colors';
import color from 'color';
import Mixins from '../../../styles/mixins';

export default {
  lightTheme: {
    ...Mixins.lightNavTheme()
  },
  darkTheme: {
    ...Mixins.darkNavTheme()
  },
  transparentTheme: {
    ...Mixins.transparentNavTheme()
  },
  lightDropdownTheme: {
    backgroundColor: '#f5f5f5',

    '& .aj-nav-list-item > div > div': {
      color: '#131313'
    }
  },
  darkDropdownTheme: {
    backgroundColor: '#131313',

    '& .aj-nav-list-item > div > div': {
      color: '#eee'
    }
  },
  lightDropdownHeaderTheme: {
    '&.sxctf-nav-bar-dropdown-visible': {
      ...Mixins.lightNavTheme()
    }
  },
  darkDropdownHeaderTheme: {
    '&.sxctf-nav-bar-dropdown-visible': {
      ...Mixins.darkNavTheme()
    }
  },
  transparentDropdownHeaderTheme: {
    '&.sxctf-nav-bar-dropdown-visible': {
      ...Mixins.transparentNavTheme()
    }
  },
  sticky: {
    position: 'fixed',
    top: 0,
    left: 0,
    right: 0,
    zIndex: 1,
    borderBottom: `2px solid ${color(Colors.spotPaletteBrown).darken(0.2)}`,

    '& .sxctf-logo': {
      height: 55,
      width: 75
    },

    '& h1': {
      fontSize: 20
    },

    '& .aj-text-button > button, & .aj-outlined-button > button, & .aj-contained-button > button': {
      fontSize: 12
    },

    '& .aj-mobile-hamburger span, & .aj-mobile-hamburger span:before, & .aj-mobile-hamburger span:after': {
      width: '25px',
      height: '3px'
    },

    '& .aj-mobile-hamburger': {
      padding: '16px 0',

      '& span:before': {
        top: '-8px'
      },

      '& span:after': {
        top: '8px'
      }
    }
  },
  dry: {
    position: 'absolute',

    '& .aj-mobile-hamburger span, & .aj-mobile-hamburger span:before, & .aj-mobile-hamburger span:after': {
      width: 30,
      height: 4
    },

    '& .aj-mobile-hamburger': {
      width: 30,
      padding: '15px 4px'
    },

    '& .aj-mobile-hamburger button span:before': {
      top: -10
    },

    '& .aj-mobile-hamburger button span:after': {
      top: 10
    }
  },
  stickyDropdown: {
    '&.sxctf-nav-dropdown-visible': {
      top: 55,
      animationName: '$show-background-sticky'
    }
  },
  dryDropdown: {
    '&.sxctf-nav-dropdown-visible': {
      top: 100,
      animationName: '$show-background'
    }
  },
  navList: {},
  '@media screen and (max-width: 450px)': {
    navList: {
      justifyContent: 'flex-start !important',
      marginTop: 50
    }
  },
  '@media screen and (max-width: 390px)': {
    dry: {
      '& .aj-mobile-hamburger span, & .aj-mobile-hamburger span:before, & .aj-mobile-hamburger span:after': {
        width: 25,
        height: 3
      },

      '& .aj-mobile-hamburger': {
        width: 25,
        padding: '10px 2px'
      },

      '& .aj-mobile-hamburger button span:before': {
        top: -8
      },

      '& .aj-mobile-hamburger button span:after': {
        top: 8
      }
    }
  },
  '@keyframes show-background': {
    from: {
      height: 0
    },
    to: {
      height: 'calc(100vh - 100px)'
    }
  },
  '@keyframes show-background-mobile': {
    from: {
      height: 0
    },
    to: {
      height: 'calc(100vh - 75px)'
    }
  },
  '@keyframes show-background-sticky': {
    from: {
      height: 0
    },
    to: {
      height: 'calc(100vh - 55px)'
    }
  }
};
