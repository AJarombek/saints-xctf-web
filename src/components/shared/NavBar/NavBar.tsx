/**
 * Home navigation bar displayed at the top of the home page.
 * @author Andrew Jarombek
 * @since 1/12/2020
 */

import React, {useState, createRef, useEffect} from 'react';
import {createUseStyles} from 'react-jss';
import { useHistory } from 'react-router-dom';
import classnames from 'classnames';
import styles from './styles';
import { AJButton, AJMobileHamburger, AJNavList } from 'jarombek-react-components';

// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
// @ts-ignore
import saintsXCTFLogo from '../../../../assets/saintsxctf_logo.png';
import {UserMeta} from '../../../redux/types';
import {useDispatch} from 'react-redux';
import {signOut, SignOutAction} from '../../../redux/modules/auth';
import {Dispatch} from "redux";

interface Props {
  includeHeaders?: Array<string>;
  user?: UserMeta;
  bodyRef: React.RefObject<any>;
}

const useStyles = createUseStyles(styles);

const handleScroll = (ref: React.RefObject<any>, setStickyHeader: (isSticky: boolean) => void): void => {
  if (ref.current) {
    setStickyHeader(ref.current.getBoundingClientRect().top <= -100);
  }
};

const NavBar: React.FunctionComponent<Props> = ({
  includeHeaders = [],
  user,
  bodyRef
}) => {
  const classes = useStyles();
  
  const dispatch = useDispatch();

  const history = useHistory();
  const [showDropdown, setShowDropdown] = useState(false);
  const [stickyHeader, setStickyHeader] = useState<boolean>(false);

  const mobileHamburgerRef: React.RefObject<HTMLInputElement> = createRef();

  const scrollEventListener = (): void => handleScroll(bodyRef, setStickyHeader);

  useEffect(() => {
    window.addEventListener('scroll', scrollEventListener);

    return (): void => {
      window.removeEventListener('scroll', scrollEventListener);
    };
  }, []);

  const stickyClass = stickyHeader ? classes.sticky : classes.dry;

  const navBarClass = showDropdown ?
    classnames('sxctf-nav-bar', 'sxctf-nav-bar-dropdown-visible', stickyClass) :
    classnames('sxctf-nav-bar', 'sxctf-nav-bar-dropdown-hidden', stickyClass);

  const dropdownClass = showDropdown ?
    classnames('sxctf-nav-dropdown', 'sxctf-nav-dropdown-visible') :
    classnames('sxctf-nav-dropdown', 'sxctf-nav-dropdown-hidden');

  /**
   * There is an issue with React Router not working with hashed routes.  This helper method uses
   * the DOM API to navigate to the hashed route.
   */
  const hashRoute = (): void => {
    const { hash } = window.location;
    if (hash !== '') {
      const id = hash.replace('#', '');
      const element = document.getElementById(id);

      if (element) {
        element.scrollIntoView();
      }
    }
  };

  /**
   * Navigate to a new page using the mobile dropdown navigation list.
   * @param path The new path to navigate to within the website.
   */
  const navigateMobile = (path: string): void => {
    setShowDropdown(!showDropdown);
    history.push(path);

    if (path.includes('#')) {
      hashRoute();
    }
  };

  const mobileNavItems = [
    {
      name: 'dashboard',
      content: 'Dashboard',
      onClick: (): void => navigateMobile('/dashboard')
    },
    {
      name: 'profile',
      content: 'Profile',
      onClick: (): void => navigateMobile(`/profile/${user?.username}`)
    },
    {
      name: 'groups',
      content: 'Groups',
      onClick: (): void => navigateMobile('/groups')
    },
    {
      name: 'admin',
      content: 'Admin',
      onClick: (): void => navigateMobile('/admin')
    },
    {
      name: 'signOut',
      content: 'Sign Out',
      onClick: (): SignOutAction => dispatch(signOut())
    },
    {
      name: 'home',
      content: 'Home',
      onClick: (): void => navigateMobile('/')
    },
    {
      name: 'about',
      content: 'About',
      onClick: (): void => {
        navigateMobile('#about');
        mobileHamburgerRef.current?.click();
      }
    },
    {
      name: 'testimonials',
      content: 'Testimonials',
      onClick: (): void => {
        navigateMobile('#testimonials');
        mobileHamburgerRef.current?.click();
      }
    },
    {
      name: 'register',
      content: 'Register',
      onClick: (): void => navigateMobile('/register')
    },
    {
      name: 'signIn',
      content: 'Sign In',
      onClick: (): void => navigateMobile('/signin')
    },
    {
      name: 'logo',
      content: <img className="mobile-dropdown-logo" src={saintsXCTFLogo} alt="" />,
      onClick: (): void => navigateMobile('#')
    }
  ];

  return (
    <>
      <div className={navBarClass}>
        <figure className="sxctf-logo">
          <img
            src={saintsXCTFLogo}
            onClick={(): void => history.push('/#')}
            alt=""
          />
        </figure>
        <h1 onClick={(): void => history.push('/#')}>SaintsXCTF</h1>
        <div className="sxctf-nav-buttons">
          { includeHeaders.includes('home') &&
            <AJButton type="text" className="homeButton" onClick={(): void => history.push('/')}>
              Home
            </AJButton>
          }
          { includeHeaders.includes('about') &&
            <AJButton type="text" className="aboutButton" onClick={(): void => {
              history.push('/#about');
              hashRoute();
            }}>
              About
            </AJButton>
          }
          { includeHeaders.includes('testimonials') &&
            <AJButton type="text"  className="testimonialsButton" onClick={(): void => {
              history.push('/#testimonials');
              hashRoute();
            }}>
              Testimonials
            </AJButton>
          }
          { includeHeaders.includes('register') &&
            <AJButton type="outlined" className="registerButton" onClick={(): void => history.push('/register')}>
              Register
            </AJButton>
          }
          { includeHeaders.includes('signIn') &&
            <AJButton type="contained" className="signInButton" onClick={(): void => history.push('/signin')}>
              Sign In
            </AJButton>
          }
          { includeHeaders.includes('dashboard') &&
            <AJButton type="text" className="dashboardButton" onClick={(): void => history.push('/dashboard')}>
              Dashboard
            </AJButton>
          }
          { includeHeaders.includes('profile') &&
            <AJButton
              type="text"
              className="profileButton"
              onClick={(): void => history.push(`/profile/${user?.username}`)}
            >
              Profile
            </AJButton>
          }
          { includeHeaders.includes('groups') &&
            <AJButton type="text" className="groupsButton" onClick={(): void => history.push('/groups')}>
              Groups
            </AJButton>
          }
          { includeHeaders.includes('admin') &&
            <AJButton type="outlined" className="adminButton" onClick={(): void => history.push('/admin')}>
              Admin
            </AJButton>
          }
          { includeHeaders.includes('signOut') &&
            <AJButton type="contained" className="signOutButton" onClick={(): SignOutAction => dispatch(signOut())}>
              Sign Out
            </AJButton>
          }
        </div>
        <div className="sxctf-nav-hamburger">
          <AJMobileHamburger
            onClick={(): void => setShowDropdown(!showDropdown)}
            ref={mobileHamburgerRef}
          />
        </div>
      </div>
      <div className={dropdownClass}>
        <AJNavList items={mobileNavItems.filter(item => includeHeaders.includes(item.name))}/>
      </div>
    </>
  );
};

export default NavBar;
