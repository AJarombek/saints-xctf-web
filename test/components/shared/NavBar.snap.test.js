/**
 * Snapshot test for the NavBar component.
 * @author Andrew Jarombek
 * @since 2/23/2020
 */

import React from 'react';
import NavBar from '../../../src/components/shared/NavBar';
import renderer from 'react-test-renderer';

// Mock react router's useHistory() hook before the tests execute.
jest.mock('react-router-dom', () => {
  const historyObj = {
    push: jest.fn()
  };

  return {
    ...jest.requireActual('react-router-dom'),
    useHistory: () => historyObj
  }
});

it('renders correctly', () => {
  const tree = renderer.create(<NavBar />).toJSON();
  expect(tree).toMatchSnapshot();
});
