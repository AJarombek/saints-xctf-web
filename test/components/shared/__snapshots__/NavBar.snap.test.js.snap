// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <div
    className="sxctf-nav-bar sxctf-nav-bar-dropdown-hidden"
  >
    <figure
      className="sxctf-logo"
    >
      <img
        alt=""
        onClick={[Function]}
        src="test-file-stub"
      />
    </figure>
    <h1
      onClick={[Function]}
    >
      SaintsXCTF
    </h1>
    <div
      className="sxctf-nav-buttons"
    />
    <div
      className="sxctf-nav-hamburger"
    >
      <div
        className="aj-mobile-hamburger ajMobileHamburger-2-2-1"
        onClick={[Function]}
      >
        <button
          type="button"
        >
          <span
            className="aj-mobile-hamburger-inactive ajMobileHamburgerInactive-2-2-3"
          />
        </button>
      </div>
    </div>
  </div>,
  <div
    className="sxctf-nav-dropdown sxctf-nav-dropdown-hidden"
  >
    <div
      className="ajNavList-2-2-4 aj-nav-list"
    />
  </div>,
]
`;
